pkg/cloud/metricsv2/metrics.go

```diff
@@ -0,0 +1,257 @@
+		}
+		return handler.HandleFinalize(ctx, input)
+	})
+}
+
+// getAttemptContextMiddleware will capture StatusCode and ErrorCode from API call attempt.
+// This will result in the StatusCode and ErrorCode captured for last attempt when publishing to metrics.
+func getAttemptContextMiddleware() middleware.FinalizeMiddleware {
+	return middleware.FinalizeMiddlewareFunc("capa/AttemptMetricContextMiddleware", func(ctx context.Context, input middleware.FinalizeInput, handler middleware.FinalizeHandler) (middleware.FinalizeOutput, middleware.Metadata, error) {
+		request := getContext(ctx)
+		if request != nil {
+			request.Attempts++
+		}
+
+		out, metadata, err := handler.HandleFinalize(ctx, input)
+
+		if request != nil {
+			if rawResp := awsmiddleware.GetRawResponse(metadata); rawResp != nil {
+				if httpResp, ok := rawResp.(*smithyhttp.Response); ok {
+					request.StatusCode = httpResp.StatusCode
+				}
+			} else {
+				request.StatusCode = -1
+			}
+
+			if err != nil {
+				smithyErr := awserrors.ParseSmithyError(err)
+				request.ErrorCode = smithyErr.ErrorCode()
+				request.StatusCode = smithyErr.StatusCode()
+			}
+		}
+
+		return out, metadata, err
+	})
+}
+
+func getRecordAWSPermissionsIssueMiddleware(target runtime.Object) middleware.FinalizeMiddleware {
+	return middleware.FinalizeMiddlewareFunc("capa/RecordAWSPermissionsIssueMiddleware", func(ctx context.Context, input middleware.FinalizeInput, handler middleware.FinalizeHandler) (middleware.FinalizeOutput, middleware.Metadata, error) {
+		output, metadata, err := handler.HandleFinalize(ctx, input)
+		if err != nil {
+			request := getContext(ctx)
+			if request != nil {
+				var errMessage string
+				smithyErr := awserrors.ParseSmithyError(err)
+				request.ErrorCode = smithyErr.ErrorCode()
+				switch request.ErrorCode {
+				case "AccessDenied", "AuthFailure", "UnauthorizedOperation", "NoCredentialProviders",
+					"ExpiredToken", "InvalidClientTokenId", "SignatureDoesNotMatch", "ValidationError":
+					record.Warnf(target, request.ErrorCode,
+						"Operation %s failed with a credentials or permission issue: %s",
+						request.OperationName,
+						errMessage,
+					)
+				}
+			}
+		}
+		return output, metadata, err
+	})
+}
+
+func initRequestContext(ctx context.Context, controller string, target runtime.Object) context.Context {
+	if middleware.GetStackValue(ctx, requestContextKey{}) == nil {
+		ctx = middleware.WithStackValue(ctx, requestContextKey{}, &RequestData{
+			Controller: controller,
+			Target:     target,
+			Attempts:   0,
+		})
+	}
+	return ctx
+}
+
+func getContext(ctx context.Context) *RequestData {
+	rctx := middleware.GetStackValue(ctx, requestContextKey{})
+	if rctx == nil {
+		return nil
+	}
+	return rctx.(*RequestData)
+}
+
+// CaptureRequestMetrics will monitor and capture request metrics.
+func (r *RequestData) CaptureRequestMetrics() {
+	if r.Service == "" || r.Region == "" || r.OperationName == "" || r.Controller == "" {
```

Comment:
Might be nice to have a `IsEmpty` or `IsIncomplete` function on `RequestData` instead.